libp2p-identify/src/protocol.rs

```diff
@@ -0,0 +1,281 @@
+#[derive(Debug, Clone)]
+pub struct IdentifyInfo {
+	/// Public key of the node in the DER format.
+	pub public_key: Vec<u8>,
+	/// Version of the "global" protocol, eg. `ipfs/1.0.0` or `polkadot/1.0.0`.
+	pub protocol_version: String,
+	/// Name and version of the client. Can be thought as similar to the `User-Agent` header
+	/// of HTTP.
+	pub agent_version: String,
+	/// Addresses that the node is listening on.
+	pub listen_addrs: Vec<Multiaddr>,
+	/// Protocols supported by the node, eg. `/ipfs/ping/1.0.0`.
+	pub protocols: Vec<String>,
+}
+
+impl<C> ConnectionUpgrade<C> for IdentifyProtocolConfig
+where
+	C: AsyncRead + AsyncWrite + 'static,
+{
+	type NamesIter = iter::Once<(Bytes, Self::UpgradeIdentifier)>;
+	type UpgradeIdentifier = ();
+	type Output = IdentifyOutput<C>;
+	type Future = Box<Future<Item = Self::Output, Error = IoError>>;
+
+	#[inline]
+	fn protocol_names(&self) -> Self::NamesIter {
+		iter::once((Bytes::from("/ipfs/id/1.0.0"), ()))
+	}
+
+	fn upgrade(self, socket: C, _: (), ty: Endpoint, observed_addr: &Multiaddr) -> Self::Future {
+		let socket = socket.framed(VarintCodec::default());
+
+		match ty {
+			Endpoint::Dialer => {
+				let future = socket
+					.into_future()
+					.map(|(msg, _)| msg)
+					.map_err(|(err, _)| err)
+					.and_then(|msg| {
+						if let Some(msg) = msg {
+							let (info, observed_addr) = parse_proto_msg(msg)?;
+							Ok(IdentifyOutput::RemoteInfo {
+								info,
+								observed_addr,
+							})
+						} else {
+							Err(IoErrorKind::InvalidData.into())
+						}
+					});
+
+				Box::new(future) as Box<_>
+			}
+
+			Endpoint::Listener => {
+				let sender = IdentifySender { inner: socket };
+
+				let future = future::ok(IdentifyOutput::Sender {
+					sender,
+					observed_addr: observed_addr.clone(),
+				});
+
+				Box::new(future) as Box<_>
+			}
+		}
+	}
+}
+
+// Turns a protobuf message into an `IdentifyInfo` and an observed address. If something bad
+// happens, turn it into an `IoError`.
+fn parse_proto_msg(msg: BytesMut) -> Result<(IdentifyInfo, Multiaddr), IoError> {
+	match protobuf_parse_from_bytes::<structs_proto::Identify>(&msg) {
+		Ok(mut msg) => {
+			let listen_addrs = {
+				let mut addrs = Vec::new();
+				for addr in msg.take_listenAddrs().into_iter() {
+					addrs.push(bytes_to_multiaddr(addr)?);
+				}
+				addrs
+			};
+
+			let observed_addr = bytes_to_multiaddr(msg.take_observedAddr())?;
+
+			let info = IdentifyInfo {
+				public_key: msg.take_publicKey(),
+				protocol_version: msg.take_protocolVersion(),
+				agent_version: msg.take_agentVersion(),
+				listen_addrs: listen_addrs,
+				protocols: msg.take_protocols().into_vec(),
+			};
+
+			Ok((info, observed_addr))
+		}
+
+		Err(err) => Err(IoError::new(IoErrorKind::InvalidData, err)),
+	}
+}
+
+// Turn a `Vec<u8>` into a `Multiaddr`. If something bad happens, turn it into an `IoError`.
+fn bytes_to_multiaddr(bytes: Vec<u8>) -> Result<Multiaddr, IoError> {
```

Comment:
This should be defined inside `parse_proto_msg` because it's very specific to that function. Better still would be to `impl Into<io::Error> for multiaddr::errors::Error` and then just do `Multiaddr::from_bytes(bytes)?` and let the `?` machinery add the `into` call.